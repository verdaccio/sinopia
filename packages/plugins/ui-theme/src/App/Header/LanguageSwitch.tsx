/* eslint-disable react/jsx-pascal-case */
import styled from '@emotion/styled';
import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
import Grid from '@mui/material/Grid';
import Typography from '@mui/material/Typography';
import i18next from 'i18next';
import React, { useCallback, useContext, useState } from 'react';
import { useTranslation } from 'react-i18next';
import ThemeContext from 'verdaccio-ui/design-tokens/ThemeContext';
import { Theme } from 'verdaccio-ui/design-tokens/theme';

import { Language, listLanguages } from '../../i18n/enabledLanguages';

const listConverted = listLanguages.reduce((prev, item) => {
  prev[item.lng] = {
    translation: item.menuKey,
    icon: item.icon,
  };
  return prev;
}, {});

export const CardSelected = styled(Card)<{ theme?: Theme }>(({ theme }) => {
  return {
    backgroundColor: theme?.palette?.grey['600'],
    opacity: '0.9',
    color: theme?.palette?.error.contrastText,
    fontWeight: 'bold',
  };
});

export const CardUnSelected = styled(Card)<{ theme?: Theme }>(({ theme }) => {
  return {
    cursor: 'pointer',
    ':hover': {
      backgroundColor: theme?.palette.greyGainsboro,
    },
  };
});

const LanguageContent = ({ translation, icon }) => (
  <>
    <CardContent>
      <Typography display="block" gutterBottom={true} variant="caption">
        {translation}
      </Typography>
    </CardContent>
    <CardContent>{icon}</CardContent>
  </>
);

const LanguageSwitch = () => {
  const themeContext = useContext(ThemeContext) as any;
  const [languages] = useState<Language[]>(
    Object.keys(i18next.options?.resources || {}) as Language[]
  );
  const { t } = useTranslation();
  const currentLanguage = themeContext.language;
  const getCurrentLngDetails = useCallback(
    (language: Language) => {
      const lng = listConverted[language] || listConverted['en-US'];
      return {
        icon: <lng.icon />,
        translation: t(lng.translation),
      };
    },
    [t]
  );
  const handleChangeLanguage = useCallback(
    (language: Language) => {
      themeContext.setLanguage(language);
    },
    [themeContext]
  );

  return (
    <div>
      <Grid columns={{ xs: 12, sm: 12, md: 12 }} container={true} spacing={{ xs: 1, md: 1 }}>
        {languages.map((language, index) => {
          const { icon, translation } = getCurrentLngDetails(language);
          return (
            <Grid item={true} key={index} sm={2} xs={6}>
              {language === currentLanguage ? (
                <CardSelected sx={{ maxWidth: 80 }}>
                  <LanguageContent icon={icon} translation={translation} />
                </CardSelected>
              ) : (
                <CardUnSelected
                  onClick={() => handleChangeLanguage(language)}
                  sx={{ maxWidth: 80 }}
                >
                  <LanguageContent icon={icon} translation={translation} />
                </CardUnSelected>
              )}
            </Grid>
          );
        })}
      </Grid>
    </div>
  );
};

export default LanguageSwitch;
