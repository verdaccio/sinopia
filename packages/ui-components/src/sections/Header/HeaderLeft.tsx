import Toolbar from '@mui/material/Toolbar';
import React from 'react';
import { Link as RouterLink } from 'react-router-dom';

import { Logo, Search } from '../../';
import { SearchWrapper } from './styles';

interface Props {
  showSearch?: boolean;
}

const HeaderLeft: React.FC<Props> = ({ showSearch }) => (
  <Toolbar
    sx={{
      display: 'flex',
      padding: 0,
      marginLeft: 1,
      flex: 1,
      '@media (min-width: 600px)': {
        padding: 0,
        marginLeft: 1,
        marginRight: '20px',
      },
    }}
  >
    <RouterLink to={'/'}>
      <Logo />
    </RouterLink>
    {showSearch && (
      <SearchWrapper data-testid="search-container">
        <Search />
      </SearchWrapper>
    )}
  </Toolbar>
);

export default HeaderLeft;
