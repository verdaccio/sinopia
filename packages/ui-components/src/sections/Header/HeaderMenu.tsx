import AccountCircle from '@mui/icons-material/AccountCircle';
import IconButton from '@mui/material/IconButton';
import Menu from '@mui/material/Menu';
import React, { MouseEvent } from 'react';
import { useTranslation } from 'react-i18next';

import { MenuItem } from '../../';
import HeaderGreetings from './HeaderGreetings';

interface Props {
  username: string;
  isMenuOpen: boolean;
  anchorEl?: Element | null | undefined;
  onLogout: () => void;
  onLoggedInMenu: (event: MouseEvent<HTMLButtonElement>) => void;
  onLoggedInMenuClose: () => void;
}

const HeaderMenu: React.FC<Props> = ({
  onLogout,
  username,
  isMenuOpen = false,
  anchorEl,
  onLoggedInMenu,
  onLoggedInMenuClose,
}) => {
  const { t } = useTranslation();
  return (
    <>
      <IconButton
        color="inherit"
        data-testid="logInDialogIcon"
        id="header--button-account"
        onClick={onLoggedInMenu}
        size="large"
      >
        <AccountCircle />
      </IconButton>
      <Menu
        anchorEl={anchorEl}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'right',
        }}
        onClose={onLoggedInMenuClose}
        open={isMenuOpen}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'right',
        }}
      >
        <MenuItem>
          <HeaderGreetings username={username} />
        </MenuItem>
        <MenuItem data-testid="logOutDialogIcon" id="logOutDialogIcon" onClick={onLogout}>
          {t('button.logout')}
        </MenuItem>
      </Menu>
    </>
  );
};

export default HeaderMenu;
