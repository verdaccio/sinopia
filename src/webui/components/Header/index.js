/**
 * @prettier
 * @flow
 */

import React, { Component } from 'react';
import type { Node } from 'react';
import { Link } from 'react-router-dom';

import Button from '@material-ui/core/Button/index';
import IconButton from '@material-ui/core/IconButton/index';
import MenuItem from '@material-ui/core/MenuItem/index';
import Menu from '@material-ui/core/Menu/index';
import Info from '@material-ui/icons/Info';
import Help from '@material-ui/icons/Help';
import Tooltip from '@material-ui/core/Tooltip/index';
import AccountCircle from '@material-ui/icons/AccountCircle';
import { default as IconSearch } from '@material-ui/icons/Search';

import { getRegistryURL } from '../../utils/url';
import ExternalLink from '../Link';
import Logo from '../Logo';
import RegistryInfoDialog from '../RegistryInfoDialog';
import Label from '../Label';
import Search from '../Search';
import RegistryInfoContent from '../RegistryInfoContent';

import { IProps, IState } from './types';
import type { ToolTipType } from './types';
import { Greetings, NavBar, InnerNavBar, MobileNavBar, InnerMobileNavBar, LeftSide, RightSide, IconSearchButton, SearchWrapper } from './styles';

class Header extends Component<IProps, IState> {
  handleLoggedInMenu: Function;
  handleLoggedInMenuClose: Function;
  handleOpenRegistryInfoDialog: Function;
  handleCloseRegistryInfoDialog: Function;
  handleToggleLogin: Function;
  renderInfoDialog: Function;

  constructor(props: IProps) {
    super(props);
    this.state = {
      openInfoDialog: false,
      registryUrl: getRegistryURL(),
      showMobileNavBar: false,
    };
  }

  /**
   * opens popover menu for logged in user.
   */
  handleLoggedInMenu = (event: SyntheticEvent<HTMLElement>) => {
    this.setState({
      anchorEl: event.currentTarget,
    });
  };

  /**
   * closes popover menu for logged in user
   */
  handleLoggedInMenuClose = () => {
    this.setState({
      anchorEl: null,
    });
  };

  /**
   * opens registry information dialog.
   */
  handleOpenRegistryInfoDialog = () => {
    this.setState({
      openInfoDialog: true,
    });
  };

  /**
   * closes registry information dialog.
   */
  handleCloseRegistryInfoDialog = () => {
    this.setState({
      openInfoDialog: false,
    });
  };

  /**
   * close/open popover menu for logged in users.
   */
  handleToggleLogin = () => {
    const { onToggleLoginModal } = this.props;
    this.setState(
      {
        anchorEl: null,
      },
      onToggleLoginModal
    );
  };

  handleToggleMNav = () => {
    const { showMobileNavBar } = this.state;
    this.setState({
      showMobileNavBar: !showMobileNavBar,
    });
  };

  handleDismissMNav = () => {
    this.setState({
      showMobileNavBar: false,
    });
  };

  renderLeftSide = (): Node => {
    const { withoutSearch = false } = this.props;
    return (
      <LeftSide>
        <Link style={{ marginRight: '1em' }} to={'/'}>
          {this.renderLogo()}
        </Link>
        {!withoutSearch && (
          <SearchWrapper>
            <Search />
          </SearchWrapper>
        )}
      </LeftSide>
    );
  };

  renderLogo = (): Node => {
    const { logo } = this.props;

    if (logo !== '') {
      return <img alt={'logo'} height={'40px'} src={logo} />;
    } else {
      return <Logo />;
    }
  };

  renderToolTipIcon = (title: string, type: ToolTipType) => {
    let content;
    switch (type) {
      case 'help':
        content = (
          <IconButton blank={true} color={'inherit'} component={ExternalLink} to={'https://verdaccio.org/docs/en/installation'}>
            <Help />
          </IconButton>
        );
        break;
      case 'info':
        content = (
          <IconButton color={'inherit'} id={'header--button-registryInfo'} onClick={this.handleOpenRegistryInfoDialog}>
            <Info />
          </IconButton>
        );
        break;
      case 'search':
        content = (
          <IconSearchButton color={'inherit'} onClick={this.handleToggleMNav}>
            <IconSearch />
          </IconSearchButton>
        );
        break;
    }
    return (
      <Tooltip disableFocusListener={true} title={title}>
        {content}
      </Tooltip>
    );
  };

  renderRightSide = (): Node => {
    const { username = '', withoutSearch = false } = this.props;
    return (
      <RightSide>
        {!withoutSearch && this.renderToolTipIcon('Search packages', 'search')}
        {this.renderToolTipIcon('Documentation', 'help')}
        {this.renderToolTipIcon('Registry Information', 'info')}
        {username ? (
          this.renderMenu()
        ) : (
          <Button color={'inherit'} id={'header--button-login'} onClick={this.handleToggleLogin}>
            {'Login'}
          </Button>
        )}
      </RightSide>
    );
  };

  renderGreetings = () => {
    const { username = '' } = this.props;
    return (
      <>
        <Greetings>{`Hi,`}</Greetings>
        <Label capitalize={true} limit={140} text={username} weight={'bold'} />
      </>
    );
  };

  /**
   * render popover menu
   */
  renderMenu = (): Node => {
    const { onLogout } = this.props;
    const { anchorEl } = this.state;
    const open = Boolean(anchorEl);
    return (
      <React.Fragment>
        <IconButton color={'inherit'} id={'header--button-account'} onClick={this.handleLoggedInMenu}>
          <AccountCircle />
        </IconButton>
        <Menu
          anchorEl={anchorEl}
          anchorOrigin={{
            vertical: 'top',
            horizontal: 'right',
          }}
          id={'sidebar-menu'}
          onClose={this.handleLoggedInMenuClose}
          open={open}
          transformOrigin={{
            vertical: 'top',
            horizontal: 'right',
          }}>
          <MenuItem disabled={true}>{this.renderGreetings()}</MenuItem>
          <MenuItem id={'header--button-logout'} onClick={onLogout}>
            {'Logout'}
          </MenuItem>
        </Menu>
      </React.Fragment>
    );
  };

  renderInfoDialog = (): Node => {
    const { scope } = this.props;
    const { openInfoDialog, registryUrl } = this.state;
    return (
      <RegistryInfoDialog onClose={this.handleCloseRegistryInfoDialog} open={openInfoDialog}>
        <RegistryInfoContent registryUrl={registryUrl} scope={scope} />
      </RegistryInfoDialog>
    );
  };

  render() {
    const { showMobileNavBar } = this.state;
    const { withoutSearch = false } = this.props;
    return (
      <div>
        <NavBar position={'static'}>
          <InnerNavBar>
            {this.renderLeftSide()}
            {this.renderRightSide()}
          </InnerNavBar>
          {this.renderInfoDialog()}
        </NavBar>
        {showMobileNavBar &&
          !withoutSearch && (
            <MobileNavBar>
              <InnerMobileNavBar>
                <Search />
              </InnerMobileNavBar>
              <Button color={'inherit'} onClick={this.handleDismissMNav}>
                {'Cancel'}
              </Button>
            </MobileNavBar>
          )}
      </div>
    );
  }
}

export default Header;
