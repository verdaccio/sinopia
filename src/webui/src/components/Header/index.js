import React from 'react';
import {Button, Dialog, Input, Alert} from 'element-react';
import isString from 'lodash/isString';
import get from 'lodash/get';
import isNumber from 'lodash/isNumber';
import {Link} from 'react-router-dom';

import API from '../../../utils/api';
import storage from '../../../utils/storage';


import classes from './header.scss';
import './logo.png';
import {getRegistryURL} from '../../../utils/url';

export default class Header extends React.Component {
  state = {
    showLogin: false,
    username: '',
    password: '',
    logo: '',
    loginError: null
  }

  constructor(props) {
    super(props);
    this.toggleLoginModal = this.toggleLoginModal.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleInput = this.handleInput.bind(this);
  }

  toggleLoginModal() {
    this.setState({
      showLogin: !this.state.showLogin
    });
    this.setState({loginError: null});
  }

  handleInput(name, e) {
    this.setState({
      [name]: e
    });
  }

  componentWillMount() {
    API.get('logo')
    .then((response) => {
      this.setState({logo: response.data});
    })
    .catch((error) => {
      throw new Error(error);
    });
  }

  async handleSubmit() {
    if (this.state.username === '' || this.state.password === '') {
      return this.setState({loginError: {
        title: 'Unable to login',
        type: 'error',
        description: 'Username or password can\'t be empty!'
      }});
    }

    try {
      let resp = await API.post(`login`, {
        data: {
          username: this.state.username,
          password: this.state.password
        }
      });

      storage.setItem('token', resp.data.token);
      storage.setItem('username', resp.data.username);
      location.reload();
    } catch (e) {
      const errorObj = {
        title: 'Unable to login',
        type: 'error'
      };
      if (get(e, 'response.status', 0) === 401) {
        errorObj.description = e.response.data.error;
      } else {
        errorObj.description = e.message;
      }
      this.setState({loginError: errorObj});
    }
  }

  get isTokenExpire() {
    const token = storage.getItem('token');

    if (!isString(token)) {
      return true;
    }

    let payload = token.split('.')[1];

    if (!payload) {
      return true;
    }

    try {
      payload = JSON.parse(atob(payload));
    } catch (err) {
      console.error('Invalid token:', err, token); // eslint-disable-line
      return false;
    }

    if (!payload.exp || !isNumber(payload.exp)) {
      return true;
    }

    const jsTimestamp = (payload.exp * 1000) - 30000; // Report as expire before (real expire time - 30s)
    const expired = Date.now() >= jsTimestamp;

    if (expired) {
      storage.clear();
    }

    return expired;
  }

  handleLogout() {
    storage.clear();
    location.reload();
  }

  renderUserActionButton() {
    if (!this.isTokenExpire) { // TODO: Check jwt token expire
      return (
        <div className={ classes.welcome }>
          Hi, {storage.getItem('username')}
          &nbsp;
          <Button type="danger" onClick={this.handleLogout}>Logout</Button>
        </div>
      );
    } else {
      return <Button type="danger" style={ {marginLeft: 'auto'} } onClick={ this.toggleLoginModal }>Login</Button>;
    }
  }

  render() {
    const registryURL = getRegistryURL();

    return (
      <header className={ classes.header }>
        <div className={ classes.headerWrap }>
          <Link to="/">
            <img src={ this.state.logo } className={ classes.logo } />
          </Link>
          <figure>
            npm set registry { registryURL }
            <br/>
            npm adduser --registry { registryURL }
          </figure>
          {this.renderUserActionButton()}
        </div>

        <Dialog
          title="Login"
          size="tiny"
          visible={ this.state.showLogin }
          onCancel={ () => this.toggleLoginModal() }
        >
          <Dialog.Body>
            { this.state.loginError &&
            <Alert
              title={this.state.loginError.title} type={this.state.loginError.type}
              description={this.state.loginError.description} showIcon={true} closable={false}>
            </Alert>
            }
            <br/>
            <Input name="username" placeholder="Username" onChange={this.handleInput('username')} />
            <br/><br/>
            <Input name="password" type="password" placeholder="Type your password" onChange={this.handleInput('password')} />
          </Dialog.Body>
          <Dialog.Footer className="dialog-footer">
            <Button onClick={ () => this.toggleLoginModal() }>
              Cancel
            </Button>
            <Button type="primary" onClick={ this.handleSubmit }>
              Login
            </Button>
          </Dialog.Footer>
        </Dialog>
      </header>
    );
  }
}
