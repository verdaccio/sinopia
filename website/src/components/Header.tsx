import React, { FunctionComponent, MouseEventHandler } from 'react';
import { Link } from 'gatsby';
import cx from 'classnames';

import { createStyles, Theme, makeStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import MenuIcon from '@material-ui/icons/Menu';

import VerdaccioWhiteLogo from './VerdaccioWhiteLogo';

export type Props = {
  onClickOpen: MouseEventHandler;
  isPermanent: boolean;
};

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    appBar: {
      // zIndex: theme.zIndex.drawer + 1,
      backgroundColor: '#FFF',
    },
    appBarShift: {
      [theme.breakpoints.up('lg')]: {
        width: 'calc(100% - 299px)',
      },
    },
    drawer: {
      [theme.breakpoints.up('lg')]: {
        flexShrink: 0,
        width: 300,
      },
    },
    navIconHide: {
      color: '#000',
      [theme.breakpoints.up('lg')]: {
        display: 'none',
      },
    },
  })
);

const Header: FunctionComponent<Props> = ({ onClickOpen, isPermanent }) => {
  // const {
  //   site: {
  //     siteMetadata: { siteName },
  //   },
  // } = useStaticQuery(graphql`
  //   query {
  //     site {
  //       siteMetadata {
  //         siteName
  //       }
  //     }
  //   }
  // `);
  const classes = useStyles();

  return (
    <AppBar
      onClick={onClickOpen}
      className={cx(classes.appBar, {
        [classes.appBarShift]: isPermanent,
      })}>
      <Toolbar>
        <IconButton
          edge="start"
          color="inherit"
          onClick={onClickOpen}
          className={classes.navIconHide}>
          <MenuIcon />
        </IconButton>
        <Typography component="h2" variant="h5">
          <Link title="Home" to="/">
            <VerdaccioWhiteLogo width="30px" />
          </Link>
        </Typography>
      </Toolbar>
    </AppBar>
  );
};

export default Header;
