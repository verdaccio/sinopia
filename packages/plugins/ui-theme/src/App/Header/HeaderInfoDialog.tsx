/* eslint-disable verdaccio/jsx-spread */
import styled from '@emotion/styled';
import Box from '@mui/material/Box';
import Tab from '@mui/material/Tab';
import Tabs from '@mui/material/Tabs';
import Typography from '@mui/material/Typography';
import React from 'react';
import { useTranslation } from 'react-i18next';
import { useSelector } from 'react-redux';
import { Theme } from 'verdaccio-ui/design-tokens/theme';

import { RootState } from '../../store/store';
import LanguageSwitch from './LanguageSwitch';
import RegistryInfoContent from './RegistryInfoContent';
import RegistryInfoDialog from './RegistryInfoDialog';

interface Props {
  isOpen: boolean;
  onCloseDialog: () => void;
}

function a11yProps(index) {
  return {
    id: `simple-tab-${index}`,
    'aria-controls': `simple-tabpanel-${index}`,
  };
}

function TabPanel(props) {
  const { children, value, index, ...other } = props;

  return (
    <div
      aria-labelledby={`simple-tab-${index}`}
      hidden={value !== index}
      id={`simple-tabpanel-${index}`}
      role="tabpanel"
      {...other}
    >
      {value === index && (
        <Box sx={{ paddingTop: 3 }}>
          <Typography>{children}</Typography>
        </Box>
      )}
    </div>
  );
}

const TextContent = styled('div')<{ theme?: Theme }>(({ theme }) => ({
  padding: '10px 0',
  backgroundColor: theme?.palette.background.default,
}));

const HeaderInfoDialog: React.FC<Props> = ({ onCloseDialog, isOpen }) => {
  const [value, setValue] = React.useState(0);

  const handleChange = (event, newValue) => {
    setValue(newValue);
  };
  const configStore = useSelector((state: RootState) => state.configuration.config);
  const { scope, base } = configStore;
  const { t } = useTranslation();
  return (
    <RegistryInfoDialog onClose={onCloseDialog} open={isOpen}>
      <Box sx={{ width: '100%' }}>
        <Box sx={{ borderBottom: 1, borderColor: 'divider' }}>
          <Tabs aria-label="basic tabs example" onChange={handleChange} value={value}>
            <Tab label={t('packageManagers.title')} {...a11yProps(0)} />
            <Tab label={t('language.title')} {...a11yProps(1)} />
          </Tabs>
        </Box>
        <TabPanel index={0} value={value}>
          <RegistryInfoContent registryUrl={base} scope={scope} />
        </TabPanel>
        <TabPanel index={1} value={value}>
          <TextContent>{t('language.description')}</TextContent>
          <LanguageSwitch />
        </TabPanel>
      </Box>
    </RegistryInfoDialog>
  );
};

export default HeaderInfoDialog;
