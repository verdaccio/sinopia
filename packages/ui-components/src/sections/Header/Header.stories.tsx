import React from 'react';
import { MemoryRouter, Route } from 'react-router';

import { HeaderInfoDialog } from '../../';
import { VersionProvider } from '../../providers';
import Header from './Header';

export default {
  title: 'Sections/Header',
};

function CustomInfoDialog({ onCloseDialog, title, isOpen }) {
  return (
    <HeaderInfoDialog
      dialogTitle={title}
      isOpen={isOpen}
      onCloseDialog={onCloseDialog}
      tabPanels={[
        { element: <div>{'foo'}</div> },
        { element: <div>{'bar'}</div> },
        { element: <div>{'fooBar'}</div> },
      ]}
      tabs={[{ label: 'foo' }, { label: 'bar' }, { label: 'barFoo' }]}
    />
  );
}

export const HeaderAll: any = () => (
  <MemoryRouter initialEntries={[`/-/web/detail/storybook`]}>
    <Route exact={true} path="/-/web/detail/:package">
      <VersionProvider>
        <Header HeaderInfoDialog={CustomInfoDialog} />
      </VersionProvider>
    </Route>
  </MemoryRouter>
);
