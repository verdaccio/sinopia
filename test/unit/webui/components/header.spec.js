/**
 * @prettier
 * @flow
 */

import React from 'react';
import {mount} from 'enzyme';
import Header from '../../../../src/webui/components/Header';

describe('<Header /> component with logged in state', () => {
  let wrapper;
  let props;

  beforeEach(() => {
    props = {
      username: 'test user',
      handleLogout: jest.fn(),
      toggleLoginModal: jest.fn(),
      scope: 'test scope',
    };
    wrapper = mount(<Header {...props} />);
  });

  test('should load the component in logged in state', () => {
    const state = {openInfoDialog: false, registryUrl: 'http://localhost'};
    expect(wrapper.state()).toEqual(state);
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('handleLoggedInMenu: set anchorEl to html element value in state', () => {
    const {handleLoggedInMenu} = wrapper.instance();

    // creates a sample menu
    const div = document.createElement('div');
    const text = document.createTextNode('sample menu');
    div.appendChild(text);

    const event = {
      currentTarget: div,
    };

    handleLoggedInMenu(event);
    expect(wrapper.state('anchorEl')).toEqual(div);
  });
});

describe('<Header /> component with logged out state', () => {
  let wrapper;
  let props;

  beforeEach(() => {
    props = {
      handleLogout: jest.fn(),
      toggleLoginModal: jest.fn(),
      scope: 'test scope',
    };
    wrapper = mount(<Header {...props} />);
  });

  test('should load the component in logged out state', () => {
    const state = {openInfoDialog: false, registryUrl: 'http://localhost'};
    expect(wrapper.state()).toEqual(state);
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('handleLoggedInMenuClose: set anchorEl value to null in state', () => {
    const {handleLoggedInMenuClose} = wrapper.instance();
    handleLoggedInMenuClose();
    expect(wrapper.state('anchorEl')).toBeNull();
  });

  test('handleOpenRegistryInfoDialog: set openInfoDialog to be truthy in state', () => {
    const {handleOpenRegistryInfoDialog} = wrapper.instance();
    handleOpenRegistryInfoDialog();
    expect(wrapper.state('openInfoDialog')).toBeTruthy();
  });

  test('handleCloseRegistryInfoDialog: set openInfoDialog to be falsy in state', () => {
    const {handleCloseRegistryInfoDialog} = wrapper.instance();
    handleCloseRegistryInfoDialog();
    expect(wrapper.state('openInfoDialog')).toBeFalsy();
  });

  test('handleToggleLogin: close/open popover menu', () => {
    const {handleToggleLogin} = wrapper.instance();
    handleToggleLogin();
    expect(wrapper.state('anchorEl')).toBeNull();
    expect(props.toggleLoginModal).toHaveBeenCalled();
  });
});
