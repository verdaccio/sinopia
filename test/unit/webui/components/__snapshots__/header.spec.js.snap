// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> component shallow should load header component in login state 1`] = `"<header class=\\"header\\"><div class=\\"headerWrap\\"><a href=\\"http://localhost/#/\\"><img src=\\"logo.png\\" class=\\"logo\\"/></a><figure>npm set scope:registry http://localhost<br/>npm adduser --registry http://localhost</figure><div class=\\"headerRight\\"><div class=\\"user-logged\\"><span class=\\"user-logged-greetings usernameField\\">Hi, Verdaccio</span><button class=\\"el-button el-button--danger headerButton header-button-logout\\" type=\\"button\\"><span>Logout</span></button></div></div></div></header>"`;

exports[`<Header /> component shallow should load header component in logout state 1`] = `"<header class=\\"header\\"><div class=\\"headerWrap\\"><a href=\\"http://localhost/#/\\"><img src=\\"logo.png\\" class=\\"logo\\"/></a><figure>npm set scope:registry http://localhost<br/>npm adduser --registry http://localhost</figure><div class=\\"headerRight\\"><button class=\\"el-button el-button--default headerButton header-button-login\\" type=\\"button\\"><span>Login</span></button></div></div></header>"`;
